staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go

```diff
@@ -70,24 +70,21 @@ const (
 
 // NewLeaderElector creates a LeaderElector from a LeaderElectionConfig
 func NewLeaderElector(lec LeaderElectionConfig) (*LeaderElector, error) {
-	if lec.LeaseDuration <= lec.RenewDeadline {
-		return nil, fmt.Errorf("leaseDuration must be greater than renewDeadline")
+	checks := []struct {
+		bad    bool
+		errMsg string
+	}{
+		{lec.LeaseDuration <= lec.RenewDeadline, "leaseDuration must be greater then renewDeadline"},
+		{lec.RenewDeadline <= time.Duration(JitterFactor*float64(lec.RetryPeriod)), "renewDeadline must be greater than retryPeriod*JitterFactor"},
+		{lec.LeaseDuration < 1, "leaseDuration must be greater than zero"},
+		{lec.RenewDeadline < 1, "renewDeadline must be greater than zero"},
+		{lec.RetryPeriod < 1, "retryPeriod must be greater than zero"},
+		{lec.Lock == nil, "Lock must not be nil."},
 	}
-	if lec.RenewDeadline <= time.Duration(JitterFactor*float64(lec.RetryPeriod)) {
-		return nil, fmt.Errorf("renewDeadline must be greater than retryPeriod*JitterFactor")
-	}
-	if lec.LeaseDuration < 1 {
-		return nil, fmt.Errorf("leaseDuration must be greater than zero")
-	}
-	if lec.RenewDeadline < 1 {
-		return nil, fmt.Errorf("renewDeadline must be greater than zero")
-	}
-	if lec.RetryPeriod < 1 {
-		return nil, fmt.Errorf("retryPeriod must be greater than zero")
-	}
-
-	if lec.Lock == nil {
-		return nil, fmt.Errorf("Lock must not be nil.")
+	for _, check := range checks {
+		if check.bad {
+			return nil, fmt.Errorf(check.errMsg)
```

Comment:
I dislike having the Errorf here. We do not *currently* have any formatted output fields in these error messages, but this precludes us from doing so. I think we need to retain that ability. You could accomplish that by changing the struct contain an error and doing the fmt.Errorf earlier. However we are then doing more work than in the current manner.

Reply:
Yes, as you shown, it precludes us from making a formatted `error` message here. It is a better choice only when all `error` messages can be constructed from a string, for formatted ones, eg, `fmt.Error("A should be greater than %d", threshold)`, we can just add code before or below this piece of code.
`if A < threshold {
    return nil, fmt.Errorf("A should be greater than %d", threshold)
}`

Reply:
> I dislike having the Errorf here. We do not _currently_ have any formatted output fields in these error messages, but this precludes us from doing so. I think we need to retain that ability. You could accomplish that by changing the struct contain an error and doing the fmt.Errorf earlier. However we are then doing more work than in the current manner.

Yes, as you shown, it precludes us from making a formatted `error` message here. It is a better choice only when all `error` messages can be constructed from a string, for formatted ones, eg, `fmt.Error("A should be greater than %d", threshold)`, we can just add code before or below this piece of code.
`if A < threshold { return nil, fmt.Errorf("A should be greater than %d", threshold) }`

Reply:
Was thinking about this and if you really care you can always call  fmt.sprintf when constructing the message.
Eg.
{lec.LeaseDuration <= lec.RenewDeadline, fmt.Sprintf("leaseDuration(%d) must be greater than renewDeadline(%d)", lec.LeaseDuration, lec.RenewDeadline)},

Reply:
> Was thinking about this and if you really care you can always call fmt.sprintf when constructing the message.
> Eg.
> {lec.LeaseDuration <= lec.RenewDeadline, fmt.Sprintf("leaseDuration(%d) must be greater than renewDeadline(%d)", lec.LeaseDuration, lec.RenewDeadline)},

Good idea.